### ISSUE TITLE
GitLab NPM Registry

### ISSUE THREAD

**Issue by alexstaroselsky** [release-it/release-it#592]
First off, thank you for your hard work in creating such an amazing and helpful tool. My question is regarding publish packages to the [GitLab NPM Registry](https://docs.gitlab.com/ee/user/packages/npm_registry/#uploading-packages). The documentation specifies that the **package.json** `publishConfig` for a given package resembles:

```json
"publishConfig": {
  "@foo:registry":"https://gitlab.com/api/v4/projects/<your_project_id>/packages/npm/"
}
```

It looks like `release-it` is specifically looking at property `publishConfig.registry` and otherwise defaults `NPM_DEFAULT_REGISTRY`. Given the `publishConfig` GitLab specifies does not have `registry` but instead `@someScope:registry`, is there any way via configurations or hooks `release-it` can support uploading packages to scoped, private GitLab NPM registries?

Thank you for any help you can provide.

**Comment by webpro** [comment]
Yes, I'll look into this shortly. Of course you/anyone is welcome to open a PR.

**Comment by webpro** [comment]
What happens if you use something like this?

```json
{
  "hooks": {
    "before:npm:init": "echo '//gitlab.com/api/v4/projects/<your_project_id>/packages/npm/:_authToken=${CI_JOB_TOKEN}'>.npmrc"
  }
}
```

Remove the `_authToken` token part depending on your situation.

Use `DEBUG=release-it:* release-it [...]` for more debug output.

**Comment by alexstaroselsky** [comment]
This can be marked as resolved. With `gitlab.release` set to `true` and `npm.publish` set to `false` in combination with setting the private scoped registry as well as adding the token for uploading to the registry to `.npmrc` prior to executing `release-it` (or with the hook described above), `publishConfig.registry` does not seem to be an issue. Thanks!

**Comment by webpro** [comment]
I stashed some initial code for this, maybe something might work in your PR. I guess in practice there is only one `@scoped:registry`? Maybe next to `registry`, should the scoped one have priority?

```js
const getScopedRegistry = publishConfig => {
  const key = Object.keys(publishConfig).find(key => key.endsWith(':registry'));
  return publishConfig[key];
};
```

**Comment by alexstaroselsky** [comment]
Something like that would definitely work. Were you thinking that this would be tied to a sub-property of `gitlab`, such as `gitlab.publish` with a respective confirmation prompt? Also do you see this as a separate module under `plugins/gitlab`? I imagine a lot of the npm plugin can be re-used with a few additional actions such as setting the [registry and tokens](https://docs.gitlab.com/ee/user/packages/npm_registry/#authenticating-with-a-ci-job-token).

**Comment by webpro** [comment]
Could it ever be desirable to publish to both the `registry` (npm or github pkg) and `@scoped:registry` (gitlab pkg)?

I imagine in practice you publish only to a single registry, in which case I think we can just use the current npm plugin.

**Comment by alexstaroselsky** [comment]
I agree that there would be no practical reason to publish both to GitLab and to NPM. The reason I bring adding a new publish module to `gitlab` are the additional steps necessary (a few lines to `.npmrc`) described in the GitLab NPM Registry documentation. Similar, but slightly different steps are needed for [Github packages](https://help.github.com/en/packages/publishing-and-managing-packages/publishing-a-package) if you choose to support those in the future.

**Comment by ksatirli** [comment]
> I imagine in practice you publish only to a single registry, in which case I think we can just use the current npm plugin.

This is interesting - I think it actually makes sense to publish packages to more than one repository. 

At a (past) project, we ran an internal version of Sonatype's Nexus, while also publishing to a private npmjs.com org. An adjacent team used Artifcatory for NPM packages and published to that one exclusively.

I believe publishing to npmjs.com was done as a "backup" (I wouldn't argue this to be the best possible approach, but it wasn't a hill worth fighting for), but there is definitely a use for having an option to publish to multiple registries.

The way we solved it back then was to simply run `release-it` with different config files. It was messy and though it worked, having an option to _"just"_ specify another registry URL would be much nicer and cleaner.

**Comment by webpro** [comment]
Looks like for both GitLab and GitHub packages the most viable way is to fiddle with the `.npmrc` file.

And it looks like a GitLab scoped registry can be set alongside a default registry, which is either an npm or GitHub registry. In release-it config:

```json
{
  "npm": {
    "publish": true,
    [...]
    "registries": [{
      "tokenRef": "NPM_TOKEN",
      "registry": "https://npm.pkg.github.com/OWNER"
    }, {
      "tokenRef": "GITLAB_TOKEN",
      "@foo:registry": "https://gitlab.com/api/v4/packages/npm/"
    }]
  },
}
```

In `package.json`:

```
"publishConfig": {
    "@foo:registry":"https://gitlab.com/api/v4/projects/<your_project_id>/packages/npm/"
  }
```

And then make sure the `.npmrc` contains the necessary information before (each) `npm publish`, e.g.:

```
//gitlab.com/api/v4/projects/<your_project_id>/packages/npm/:_authToken=${GITLAB_TOKEN}
//gitlab.com/api/v4/packages/npm/:_authToken=${GITLAB_TOKEN}
@foo:registry=https://gitlab.com/api/v4/packages/npm/
```

And perhaps eventually just put back the original `.npmrc` file if there was any.

**Comment by ksatirli** [comment]
Thanks for the in-depth reply - this is super helpful!

I feared that `.npmrc` was the _correct_ route - it just doesn't _feel_ like the right way.

**Comment by alexstaroselsky** [comment]
There would still need to be changes to the npm module? No matter what, it is looking for `publishConfig.registry` as that is what is passed to final `npm publish` command via `--registry` and uses that value for logging/notifications right?

**Comment by webpro** [comment]
Yes, the release-it config I showed above was just an example of how this could look like.

Most of the npm plugin would need to do a full run once for each registry, basically everything except for the bump (`npm version`), this should be done only once. Example integration test that shows a full run: https://github.com/release-it/release-it/blob/master/test/tasks.js#L178-L183

**Comment by webpro** [comment]
If only I could test this against a real gitlab registry... It's only included in their premium version. Anybody has ideas? The community/enterprise editions (docker) don't seem to include this.

**Comment by alexstaroselsky** [comment]
I have access to a private enterprise gitlab registry. I could help to test any proposed changes.

**Comment by webpro** [comment]
Thanks @alexstaroselsky!

For now I would like to focus on a single registry. I'm afraid publishing to multiple registries will be significantly more work, but we'll see about that later.

Given this (passing) test:

```
test('should publish to a different/scoped registry', async t => {
  delete require.cache[require.resolve('../package.json')];
  mock({
    [path.resolve('package.json')]: JSON.stringify({
      name: '@my-scope/my-pkg',
      version: '1.0.0',
      publishConfig: {
        '@my-scope:registry': 'https://gitlab.com/api/v4/projects/my-scope%2Fmy-pkg/packages/npm/'
      }
    })
  });
  const options = { npm };
  const npmClient = factory(npm, { options });
  const exec = sinon.stub(npmClient.shell, 'exec').resolves();

  await runTasks(npmClient);

  t.deepEqual(getNpmArgs(exec.args), [
    'npm ping --registry https://gitlab.com/api/v4/projects/my-scope%2Fmy-pkg/packages/npm/',
    'npm whoami --registry https://gitlab.com/api/v4/projects/my-scope%2Fmy-pkg/packages/npm/',
    'npm show @my-scope/my-pkg@latest version --registry https://gitlab.com/api/v4/projects/my-scope%2Fmy-pkg/packages/npm/',
    'npm version 1.0.1 --no-git-tag-version',
    'npm publish . --tag latest --access public'
  ]);

  exec.restore();
});
```

See the `@my-scope:registry` in the mocked `package.json` and the `npm` commands.

* Could you please run the three ping/whoami/show commands against your GitLab npm registry?
* I guess with your custom `.npmrc`, the `npm publish` should also work?

I'm happy to publish this version of `release-it` with some `next` tag or something, just entering the waters here.

**Comment by alexstaroselsky** [comment]
Publishing with some sort of tag such as next, beta, or alpha would definitely make it easier.

**Comment by webpro** [comment]
Of course, it has the `try` tag and you can install using e.g. `npm install -D release-it@try`

**Comment by alexstaroselsky** [comment]
I'm still working through setting this up in our GitLab CI process, but I was able to successfully execute commands locally including publishing packages to our private GitLab NPM Registry including pre-release versions. Examples of commands I tried and seemed to work:

```bash
GITLAB_TOKEN=foobar npx release-it@try --ci --preRelease=beta --git.commitMessage="chore(release): Release ${version} [skip-ci]"
```

```bash
GITLAB_TOKEN=foobar npx release-it@try --ci --git.commitMessage="chore(release): Release ${version} [skip-ci]"
```

This assumed an .npmrc with the following content:

```
@foo:registry=https://gitlab.com/api/v4/packages/npm/
//gitlab.com/api/v4/projects/12345678/packages/npm/:_authToken=foobar
//gitlab.com/api/v4/packages/npm/:_authToken=foobar
```

Just in case here is the debug output:

```
npx: installed 307 in 11.201s
  release-it:config { cosmiconfig: null } +0ms
  release-it:config { system:
  release-it:config    { 'release-it': '13.2.0-try.0',
  release-it:config      node: 'v10.17.0',
  release-it:config      os: 'macOS Catalina' } } +4ms
  release-it:config { _: [],
  release-it:config   ci: true,
  release-it:config   preRelease: 'beta',
  release-it:config   git:
  release-it:config    { commitMessage: 'chore(release): Release  [skip-ci]',
  release-it:config      changelog: 'git log --pretty=format:"* +1ms (%h)" ${latestTag}...HEAD',
  release-it:config      requireCleanWorkingDir: true,
  release-it:config      requireBranch: false,
  release-it:config      requireUpstream: true,
  release-it:config      requireCommits: false,
  release-it:config      addUntrackedFiles: false,
  release-it:config      commit: true,
  release-it:config      commitArgs: [],
  release-it:config      tag: true,
  release-it:config      tagName: null,
  release-it:config      tagAnnotation: 'Release ${version}',
  release-it:config      tagArgs: [],
  release-it:config      push: true,
  release-it:config      pushArgs: [ '--follow-tags' ],
  release-it:config      pushRepo: '' },
  release-it:config   'dry-run': false,
  release-it:config   d: false,
  release-it:config   dryRun: false,
  release-it:config   verbose: 0,
  release-it:config   V: 0,
  release-it:config   increment: undefined,
  release-it:config   hooks: {},
  release-it:config   npm:
  release-it:config    { publish: true,
  release-it:config      publishPath: '.',
  release-it:config      tag: null,
  release-it:config      access: null,
  release-it:config      otp: null,
  release-it:config      ignoreVersion: false },
  release-it:config   github:
  release-it:config    { release: false,
  release-it:config      releaseName: 'Release ${version}',
  release-it:config      releaseNotes: null,
  release-it:config      preRelease: false,
  release-it:config      draft: false,
  release-it:config      tokenRef: 'GITHUB_TOKEN',
  release-it:config      assets: null,
  release-it:config      host: null,
  release-it:config      timeout: 0,
  release-it:config      proxy: null },
  release-it:config   gitlab:
  release-it:config    { release: false,
  release-it:config      releaseName: 'Release ${version}',
  release-it:config      releaseNotes: null,
  release-it:config      tokenRef: 'GITLAB_TOKEN',
  release-it:config      assets: null,
  release-it:config      origin: null },
  release-it:config   version:
  release-it:config    { increment: undefined,
  release-it:config      isPreRelease: true,
  release-it:config      preReleaseId: 'beta' } }
  release-it:plugins { namespace: 'version',
  release-it:plugins   options:
  release-it:plugins    { increment: undefined,
  release-it:plugins      isPreRelease: true,
  release-it:plugins      preReleaseId: 'beta' } } +0ms
  release-it:plugins { namespace: 'npm',
  release-it:plugins   options:
  release-it:plugins    { publish: true,
  release-it:plugins      publishPath: '.',
  release-it:plugins      tag: null,
  release-it:plugins      access: null,
  release-it:plugins      otp: null,
  release-it:plugins      ignoreVersion: false } } +0ms
  release-it:plugins { namespace: 'git',
  release-it:plugins   options:
  release-it:plugins    { commitMessage: 'chore(release): Release  [skip-ci]',
  release-it:plugins      changelog: 'git log --pretty=format:"* +13ms (%h)" ${latestTag}...HEAD',
  release-it:plugins      requireCleanWorkingDir: true,
  release-it:plugins      requireBranch: false,
  release-it:plugins      requireUpstream: true,
  release-it:plugins      requireCommits: false,
  release-it:plugins      addUntrackedFiles: false,
  release-it:plugins      commit: true,
  release-it:plugins      commitArgs: [],
  release-it:plugins      tag: true,
  release-it:plugins      tagName: null,
  release-it:plugins      tagAnnotation: 'Release ${version}',
  release-it:plugins      tagArgs: [],
  release-it:plugins      push: true,
  release-it:plugins      pushArgs: [ '--follow-tags' ],
  release-it:plugins      pushRepo: '' } }
  release-it:metrics { url: 'http://www.google-analytics.com/collect',
  release-it:metrics   statusCode: 200,
  release-it:metrics   statusMessage: 'OK',
  release-it:metrics   payload:
  release-it:metrics    'v=1&tid=UA-108828841-1&cid=4ddc3d52-eaa2-4722-8828-7bc16082cbc5&vp=168x16&sd=1-bit&cd1=13.2.0-try.0&cd2=v10.17.0&cd3=macOS+Catalina&cd4=1&cd5=0&cd6=0&cd7=0&cd8=null&cd9=undefined&cd11=0&cd12=1&cd13=1&cd14=0&cd15=undefined&cd16=0&t=event&ec=session&ea=start' } +0ms
npm notice PING https://gitlab.com/api/v4/projects/12345678/packages/npm/
npm ERR! code E404
npm ERR! 404 Not Found - GET https://gitlab.com/api/v4/projects/12345678/packages/npm/-/ping?write=true - 404 Not Found
npm ERR! 404 
npm ERR! 404  'ping' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404 
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.
npm ERR! code E404
npm ERR! 404 Not Found - GET https://gitlab.com/api/v4/projects/12345678/packages/npm/-/whoami - 404 Not Found
npm ERR! 404 
npm ERR! 404  'whoami' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404 
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/SomeUser/.npm/_logs/2020-03-22T21_15_06_598Z-debug.log
  release-it:shell { command:
  release-it:shell    'npm ping --registry https://gitlab.com/api/v4/projects/12345678/packages/npm/',
  release-it:shell   options: {},
  release-it:shell   code: 1,
  release-it:shell   stdout: '',
  release-it:shell   stderr:
  release-it:shell    'npm notice PING https://gitlab.com/api/v4/projects/12345678/packages/npm/\nnpm ERR! code E404\nnpm ERR! 404 Not Found - GET https://gitlab.com/api/v4/projects/12345678/packages/npm/-/ping?write=true - 404 Not Found\nnpm ERR! 404 \nnpm ERR! 404  \'ping\' is not in the npm registry.\nnpm ERR! 404 You should bug the author to publish it (or use the name yourself!)\nnpm ERR! 404 \nnpm ERR! 404 Note that you can also install from a\nnpm ERR! 404 tarball, folder, http url, or git url.\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /Users/SomeUser/.npm/_logs/2020-03-22T21_15_06_598Z-debug.log\n' } +0ms
WARNING Ignoring unsupported `npm ping` command response.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/SomeUser/.npm/_logs/2020-03-22T21_15_06_609Z-debug.log
  release-it:shell { command:
  release-it:shell    'npm whoami --registry https://gitlab.com/api/v4/projects/12345678/packages/npm/',
  release-it:shell   options: {},
  release-it:shell   code: 1,
  release-it:shell   stdout: '',
  release-it:shell   stderr:
  release-it:shell    'npm ERR! code E404\nnpm ERR! 404 Not Found - GET https://gitlab.com/api/v4/projects/12345678/packages/npm/-/whoami - 404 Not Found\nnpm ERR! 404 \nnpm ERR! 404  \'whoami\' is not in the npm registry.\nnpm ERR! 404 You should bug the author to publish it (or use the name yourself!)\nnpm ERR! 404 \nnpm ERR! 404 Note that you can also install from a\nnpm ERR! 404 tarball, folder, http url, or git url.\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /Users/SomeUser/.npm/_logs/2020-03-22T21_15_06_609Z-debug.log\n' } +9ms
  release-it:npm Error: npm ERR! code E404
  release-it:npm npm ERR! 404 Not Found - GET https://gitlab.com/api/v4/projects/12345678/packages/npm/-/whoami - 404 Not Found
  release-it:npm npm ERR! 404 
  release-it:npm npm ERR! 404  'whoami' is not in the npm registry.
  release-it:npm npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
  release-it:npm npm ERR! 404 
  release-it:npm npm ERR! 404 Note that you can also install from a
  release-it:npm npm ERR! 404 tarball, folder, http url, or git url.
  release-it:npm 
  release-it:npm npm ERR! A complete log of this run can be found in:
  release-it:npm npm ERR!     /Users/SomeUser/.npm/_logs/2020-03-22T21_15_06_609Z-debug.log
  release-it:npm 
  release-it:npm     at sh.exec (/Users/SomeUser/.npm/_npx/5668/lib/node_modules/release-it/lib/shell.js:55:18)
  release-it:npm     at /Users/SomeUser/.npm/_npx/5668/lib/node_modules/release-it/node_modules/shelljs/src/exec.js:127:9
  release-it:npm     at ChildProcess.exithandler (child_process.js:301:5)
  release-it:npm     at ChildProcess.emit (events.js:198:13)
  release-it:npm     at maybeClose (internal/child_process.js:982:16)
  release-it:npm     at Socket.stream.socket.on (internal/child_process.js:389:11)
  release-it:npm     at Socket.emit (events.js:198:13)
  release-it:npm     at Pipe._handle.close (net.js:607:12) +0ms
WARNING Ignoring unsupported `npm whoami` command response.
1.0.2-beta.1
  release-it:shell { command:
  release-it:shell    'npm show @foo/test-module@beta version --registry https://gitlab.com/api/v4/projects/12345678/packages/npm/',
  release-it:shell   options: { write: false },
  release-it:shell   code: 0,
  release-it:shell   stdout: '1.0.2-beta.1',
  release-it:shell   stderr: '' } +221ms
  release-it:shell { command: 'git diff-index --quiet HEAD --',
  release-it:shell   options: { write: false },
  release-it:shell   code: 0,
  release-it:shell   stdout: '',
  release-it:shell   stderr: '' } +19ms
feat/some-branch
  release-it:shell { command: 'git rev-parse --abbrev-ref HEAD',
  release-it:shell   options: { write: false },
  release-it:shell   code: 0,
  release-it:shell   stdout: 'feat/some-branch',
  release-it:shell   stderr: '' } +20ms
origin
  release-it:shell { command: 'git config --get branch.feat/some-branch.remote',
  release-it:shell   options: { write: false },
  release-it:shell   code: 0,
  release-it:shell   stdout: 'origin',
  release-it:shell   stderr: '' } +18ms
git@gitlab.com:foo/bar/baz/libraries/test-module.git
  release-it:shell { command: 'git config --get remote.origin.url',
  release-it:shell   options: { write: false },
  release-it:shell   code: 0,
  release-it:shell   stdout:
  release-it:shell    'git@gitlab.com:foo/bar/baz/libraries/test-module.git',
  release-it:shell   stderr: '' } +16ms
  release-it:shell { command: 'git fetch',
  release-it:shell   options: {},
  release-it:shell   code: 0,
  release-it:shell   stdout: '',
  release-it:shell   stderr: '' } +1s
1.0.2-beta.1
  release-it:shell { command: 'git describe --tags --abbrev=0',
  release-it:shell   options: { write: false },
  release-it:shell   code: 0,
  release-it:shell   stdout: '1.0.2-beta.1',
  release-it:shell   stderr: '' } +26ms
  release-it:shell { command: 'git log --pretty=format:"* +23ms (%h)" 1.0.2-beta.1...HEAD',
  release-it:shell   options: { write: false },
  release-it:shell   code: 0,
  release-it:shell   stdout: '',
  release-it:shell   stderr: '' }
refs/heads/feat/some-branch
  release-it:shell { command: 'git symbolic-ref HEAD',
  release-it:shell   options: { write: false },
  release-it:shell   code: 0,
  release-it:shell   stdout: 'refs/heads/feat/some-branch',
  release-it:shell   stderr: '' } +17ms
origin/feat/some-branch
  release-it:shell { command:
  release-it:shell    'git for-each-ref --format="%(upstream:short)" refs/heads/feat/some-branch',
  release-it:shell   options: { write: false },
  release-it:shell   code: 0,
  release-it:shell   stdout: 'origin/feat/some-branch',
  release-it:shell   stderr: '' } +18ms
  release-it:config { name: '@foo/test-module',
  release-it:config   latestVersion: '1.0.2-beta.1',
  release-it:config   version: '1.0.2-beta.2',
  release-it:config   repo:
  release-it:config    { host: 'gitlab.com',
  release-it:config      owner: 'foo/bar/baz/libraries',
  release-it:config      project: 'test-module',
  release-it:config      protocol: 'ssh',
  release-it:config      remote:
  release-it:config       'git@gitlab.com:foo/bar/baz/libraries/test-module.git',
  release-it:config      repository: 'foo/bar/baz/libraries/test-module' },
  release-it:config   changelog: undefined } +3s
🚀 Let's release @foo/test-module (1.0.2-beta.1...1.0.2-beta.2)
Empty changelog
  release-it:config { version: '1.0.2-beta.2',
  release-it:config   isPreRelease: true,
  release-it:config   preReleaseId: 'beta' } +11ms
v1.0.2-beta.2
  release-it:shell { command: 'npm version 1.0.2-beta.2 --no-git-tag-version',
  release-it:shell   options: {},
  release-it:shell   code: 0,
  release-it:shell   stdout: 'v1.0.2-beta.2',
  release-it:shell   stderr: '' } +368ms
 M package-lock.json
 M package.json
  release-it:shell { command: 'git status --short --untracked-files=no',
  release-it:shell   options: { write: false },
  release-it:shell   code: 0,
  release-it:shell   stdout: 'M package-lock.json\n M package.json',
  release-it:shell   stderr: '' } +19ms
Changeset:
M package-lock.json
 M package.json
  release-it:shell { command: [ 'git', 'add', '.', '--update' ],
  release-it:shell   options: {},
  release-it:shell   stdout: '',
  release-it:shell   stderr: '' } +19ms
  release-it:shell { command:
  release-it:shell    [ 'git',
  release-it:shell      'commit',
  release-it:shell      '--message',
  release-it:shell      'chore(release): Release  [skip-ci]' ],
  release-it:shell   options: {},
  release-it:shell   stdout:
  release-it:shell    '[feat/some-branch ff4f43e] chore(release): Release  [skip-ci]\n 2 files changed, 2 insertions(+), 2 deletions(-)',
  release-it:shell   stderr: '' } +30ms
  release-it:shell { command:
  release-it:shell    [ 'git',
  release-it:shell      'tag',
  release-it:shell      '--annotate',
  release-it:shell      '--message',
  release-it:shell      'Release 1.0.2-beta.2',
  release-it:shell      '1.0.2-beta.2' ],
  release-it:shell   options: {},
  release-it:shell   stdout: '',
  release-it:shell   stderr: '' } +19ms
refs/heads/feat/some-branch
  release-it:shell { command: 'git symbolic-ref HEAD',
  release-it:shell   options: { write: false },
  release-it:shell   code: 0,
  release-it:shell   stdout: 'refs/heads/feat/some-branch',
  release-it:shell   stderr: '' } +18ms
origin/feat/some-branch
  release-it:shell { command:
  release-it:shell    'git for-each-ref --format="%(upstream:short)" refs/heads/feat/some-branch',
  release-it:shell   options: { write: false },
  release-it:shell   code: 0,
  release-it:shell   stdout: 'origin/feat/some-branch',
  release-it:shell   stderr: '' } +17ms
  release-it:shell { command: [ 'git', 'push', '--follow-tags' ],
  release-it:shell   options: {},
  release-it:shell   stdout: '',
  release-it:shell   stderr:
  release-it:shell    'remote: \nremote: View merge request for feat/some-branch:        \nremote:   https://gitlab.com/foo/bar/baz/libraries/test-module/-/merge_requests/1        \nremote: \nTo gitlab.com:foo/bar/baz/libraries/test-module.git\n   e95804f..ff4f43e  feat/some-branch -> feat/some-branch\n * [new tag]         1.0.2-beta.2 -> 1.0.2-beta.2' } +2s
npm notice 
npm notice 📦  @foo/test-module@1.0.2-beta.2
npm notice === Tarball Contents === 
npm notice 131B  .editorconfig                                              
npm notice 131B  foo/node_modules/@foo/test-module/.editorconfig 
npm notice 109B  foo/node_modules/@foo/test-module/index.js      
npm notice 109B  index.js                                                   
npm notice 381B  foo/package-lock.json                                      
npm notice 1.5kB foo/node_modules/@foo/test-module/package.json  
npm notice 286B  foo/package.json                                           
npm notice 525B  package.json                                               
npm notice 30B   foo/node_modules/@foo/test-module/README.md     
npm notice 30B   README.md                                                  
npm notice 159B  .gitlab-ci.yml                                             
npm notice 159B  foo/node_modules/@foo/test-module/.gitlab-ci.yml
npm notice === Tarball Details === 
npm notice name:          @foo/test-module             
npm notice version:       1.0.2-beta.2                            
npm notice package size:  1.4 kB                                  
npm notice unpacked size: 3.5 kB                                  
npm notice shasum:        2b50264c5e1cb4a150d84b142867d6c710b975ed
npm notice integrity:     sha512-Xc9j9T2qdLSij[...]9A/xyEb2XfySQ==
npm notice total files:   12                                      
npm notice 
+ @foo/test-module@1.0.2-beta.2
  release-it:shell { command: 'npm publish . --tag beta   ',
  release-it:shell   options: { write: false },
  release-it:shell   code: 0,
  release-it:shell   stdout: '+ @foo/test-module@1.0.2-beta.2',
  release-it:shell   stderr:
  release-it:shell    'npm notice \nnpm notice 📦  @foo/test-module@1.0.2-beta.2\nnpm notice === Tarball Contents === \nnpm notice 131B  .editorconfig                                              \nnpm notice 131B  foo/node_modules/@foo/test-module/.editorconfig \nnpm notice 109B  foo/node_modules/@foo/test-module/index.js      \nnpm notice 109B  index.js                                                   \nnpm notice 381B  foo/package-lock.json                                      \nnpm notice 1.5kB foo/node_modules/@foo/test-module/package.json  \nnpm notice 286B  foo/package.json                                           \nnpm notice 525B  package.json                                               \nnpm notice 30B   foo/node_modules/@foo/test-module/README.md     \nnpm notice 30B   README.md                                                  \nnpm notice 159B  .gitlab-ci.yml                                             \nnpm notice 159B  foo/node_modules/@foo/test-module/.gitlab-ci.yml\nnpm notice === Tarball Details === \nnpm notice name:          @foo/test-module             \nnpm notice version:       1.0.2-beta.2                            \nnpm notice package size:  1.4 kB                                  \nnpm notice unpacked size: 3.5 kB                                  \nnpm notice shasum:        2b50264c5e1cb4a150d84b142867d6c710b975ed\nnpm notice integrity:     sha512-Xc9j9T2qdLSij[...]9A/xyEb2XfySQ==\nnpm notice total files:   12                                      \nnpm notice \n' } +1s
🔗 https://gitlab.com/api/v4/projects/12345678/packages/npm/package/@foo/test-module
  release-it:metrics { url: 'http://www.google-analytics.com/collect',
  release-it:metrics   statusCode: 200,
  release-it:metrics   statusMessage: 'OK',
  release-it:metrics   payload: 't=event&ec=session&ea=end' } +7s
🏁 Done (in 18s.)
```

I'll share more when I've tried this in the CI process successfully.

Thanks!

**Comment by webpro** [comment]
Thanks, very helpful. And promising 👍

**Comment by webpro** [comment]
I've released [v13.6.1](https://github.com/release-it/release-it/releases/tag/13.6.1), which should resolve the original issue. Feel free to open a new ticket for further requests/discussion!

**Comment by ebekebe** [comment]
I noticed that release-it v13.6.2 doesn't seem to honor the environment variables `NPM_REGISTRY` and `NPM_TOKEN`. We use it to publish to the github registry without the need to touch config files like `.npmrc`. Is that something you are aware of and decided not to do or would that be a feasible addition? Currently, we need to manually run `npm publish` after running `release-it`.

**Comment by webpro** [comment]
@ebekebe release-it just executes `npm publish`, so it's `npm` that should have access to and honor those environment variables. Are you having this issue in GitLab CI environment? A simple "proof" that environment variables are available and should work:

```
❯ export FOO=BAR
❯ npm run release -- --verbose --hooks.before:init='echo \$FOO'

> release-it@13.6.3 release /Users/lars/Projects/release-it
> ./bin/release-it.js "--verbose" "--hooks.before:init=echo \$FOO"

$ echo $FOO
BAR
```

Please open a new ticket if you are having issues with this, as this ticket was already closed and your issue seems to be different.
